<?xml version="1.0"?>
<launch>
  <arg name="world_name"/>
  <arg name="name"/>
  <arg name="link_name" value="base_link"/>
  <arg name="sensor_prefix" value="/world/$(arg world_name)/model/$(arg name)/link/$(arg link_name)/sensor"/>
  <arg name="uav" default="0"/>
  <arg name="stereo_cam" default="0"/>
  <arg name="laser_scan" default="1"/>
  <arg name="rgbd_cam" default="0"/>
  <arg name="lidar_3d" default="0"/>
  <group ns="$(arg name)">
    <!--Create multipe bridges so that it can run in parallel-->
    <node
      pkg="ros1_ign_bridge"
      type="parameter_bridge"
      name="ros1_ign_bridge_imu"
      args="$(arg sensor_prefix)/imu_sensor/imu@sensor_msgs/Imu@ignition.msgs.IMU">
      <remap from="$(arg sensor_prefix)/imu_sensor/imu" to="imu/data"/>
    </node>
    <node
      pkg="ros1_ign_bridge"
      type="parameter_bridge"
      name="ros1_ign_bridge_pose"
      args="/model/$(arg name)/pose@geometry_msgs/TransformStamped@ignition.msgs.Pose">
      <remap from="/model/$(arg name)/pose" to="pose"/>
    </node>
    <node
      pkg="ros1_ign_bridge"
      type="parameter_bridge"
      name="ros1_ign_bridge_twist"
      args="/model/$(arg name)/cmd_vel_relay@geometry_msgs/Twist@ignition.msgs.Twist">
      <remap from="/model/$(arg name)/cmd_vel_relay" to="cmd_vel"/>
    </node>

    <!-- RGBD -->
    <group if="$(arg rgbd_cam)">
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_rgbd"
        args="$(arg sensor_prefix)/camera_front/points@sensor_msgs/PointCloud2@ignition.msgs.PointCloudPacked">
        <remap from="$(arg sensor_prefix)/camera_front/points" to="rgbd_camera/depth/points"/>
      </node>
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_depth"
        args="$(arg sensor_prefix)/camera_front/depth_image@sensor_msgs/Image@ignition.msgs.Image">
        <remap from="$(arg sensor_prefix)/camera_front/depth_image" to="front/depth"/>
      </node>
    </group>

    <!-- 3D lidar -->
    <group if="$(arg lidar_3d)">
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_lidar3d"
        args="$(arg sensor_prefix)/front_laser/scan/points@sensor_msgs/PointCloud2@ignition.msgs.PointCloudPacked">
        <remap from="$(arg sensor_prefix)/front_laser/scan/points" to="points"/>
      </node>
    </group>

    <!-- stereo -->
    <group if="$(arg stereo_cam)">
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_camera"
        args="$(arg sensor_prefix)/camera_front_right/image@sensor_msgs/Image@ignition.msgs.Image
              $(arg sensor_prefix)/camera_front_right/camera_info@sensor_msgs/CameraInfo@ignition.msgs.CameraInfo
              $(arg sensor_prefix)/camera_front_left/image@sensor_msgs/Image@ignition.msgs.Image
              $(arg sensor_prefix)/camera_front_left/camera_info@sensor_msgs/CameraInfo@ignition.msgs.CameraInfo">
        <remap from="$(arg sensor_prefix)/camera_front_right/image" to="front/right/image_raw"/>
        <remap from="$(arg sensor_prefix)/camera_front_right/camera_info" to="front/right/camera_info"/>
        <remap from="$(arg sensor_prefix)/camera_front_left/image" to="front/left/image_raw"/>
        <remap from="$(arg sensor_prefix)/camera_front_left/camera_info" to="front/left/camera_info"/>
      </node>
    </group>

    <!-- non-stereo -->
    <group unless="$(arg stereo_cam)">
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_camera"
        args="$(arg sensor_prefix)/camera_front/image@sensor_msgs/Image@ignition.msgs.Image
              $(arg sensor_prefix)/camera_front/camera_info@sensor_msgs/CameraInfo@ignition.msgs.CameraInfo">
        <remap from="$(arg sensor_prefix)/camera_front/image" to="front/image_raw"/>
        <remap from="$(arg sensor_prefix)/camera_front/camera_info" to="front/camera_info"/>
      </node>
    </group>

    <!-- laser scan -->
    <group if="$(arg laser_scan)">
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_gpu_lidar"
        args="$(arg sensor_prefix)/front_laser/scan@sensor_msgs/LaserScan@ignition.msgs.LaserScan">
        <remap from="$(arg sensor_prefix)/front_laser/scan" to="front_scan"/>
      </node>
    </group>

    <!-- UAV -->
    <group if="$(arg uav)">
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_actuators"
        args="/model/$(arg name)/command/motor_speed@mav_msgs/Actuators@ignition.msgs.Actuators">
        <remap from="/model/$(arg name)/command/motor_speed" to="command/motor_speed"/>
      </node>
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_magnetometer"
        args="$(arg sensor_prefix)/magnetometer/magnetometer@sensor_msgs/MagneticField@ignition.msgs.Magnetometer">
        <remap from="$(arg sensor_prefix)/magnetometer/magnetometer" to="magnetic_field"/>
      </node>
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_fluid_pressure"
        args="$(arg sensor_prefix)/air_pressure/air_pressure@sensor_msgs/FluidPressure@ignition.msgs.FluidPressure">
        <remap from="$(arg sensor_prefix)/air_pressure/air_pressure" to="air_pressure"/>
      </node>
    </group>

    <!-- UGV -->
    <group unless="$(arg uav)">
      <node
        pkg="ros1_ign_bridge"
        type="parameter_bridge"
        name="ros1_ign_bridge_odom"
        args="/model/$(arg name)/odometry@nav_msgs/Odometry@ignition.msgs.Odometry">
        <remap from="/model/$(arg name)/odometry" to="odom"/>
      </node>
    </group>

    <node
      pkg="subt_ros"
      type="pose_tf_broadcaster"
      name="pose_tf_broadcaster"/>
  </group>
</launch>
